File: transaction.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "utils.h"
#include "account.h"

void depositMoney() {
    char accountName[50];
    printf("Entrez le nom du compte : ");
    scanf("%s", accountName);

    // Demander la catégorie d'objet pour le dépôt
    char* depositCategories[] = {"Argent", "CarteBancaire", "Permis", "PhotosIdentite", "CartesVisite"};
    int depositCategory;
    printf("Sélectionnez la catégorie pour le dépôt : \n");
    for (int i = 0; i < sizeof(depositCategories) / sizeof(depositCategories[0]); i++) {
        printf("%d. %s\n", i + 1, depositCategories[i]);
    }
    scanf("%d", &depositCategory);
    depositCategory--;

    if (depositCategory == -1) {
        printf("Opération annulée.\n");
        return;
    }

    double depositAmount = 0;
    Balance currentBalance = getCurrentBalance(accountName);

    switch (depositCategory) {
        case 0: // Argent
            double currentArgent = currentBalance.argent;
            double limitArgent = getCategoryLimit("Argent");
            printf("Entrez le montant de Argent à déposer (limite : %.2lf) : ", limitArgent - currentArgent);
            scanf("%lf", &depositAmount);

            if (currentArgent + depositAmount > limitArgent) {
                printf("Le montant total de Argent dépasse la limite de %.2lf. Opération annulée.\n", limitArgent);
                return;
            }
            break;
        case 1: // CarteBancaire
            int currentCarte = currentBalance.cartesBancaires;
            int limitCarte = getCategoryLimit("CarteBancaire");
            printf("Entrez le nombre de carte bancaire à déposer (limite : %d) : ", limitCarte - currentCarte);
            scanf("%d", &depositAmount);

            if (currentCarte + depositAmount > limitCarte) {
                printf("Le nombre dépasse la limite de %d. Opération annulée.\n", limitCarte);
                return;
            }
            break;
        case 2: // Permis
            int currentPermis = currentBalance.permisDeConduire;
            int limitPermis = getCategoryLimit("Permis");
            printf("Entrez le nombre de permis à déposer (limite : %d) : ", limitPermis - currentPermis);
            scanf("%d", &depositAmount);

            if (currentPermis + depositAmount > limitPermis) {
                printf("Le nombre dépasse la limite de %d. Opération annulée.\n", limitPermis);
                return;
            }
            break;
        case 3: // PhotosIdentite
            int currentPhoto = currentBalance.photosIdentite;
            int limitPhoto = getCategoryLimit("PhotosIdentite");
            printf("Entrez le nombre de photo à déposer (limite : %d) : ", limitPhoto - currentPhoto);
            scanf("%d", &depositAmount);

            if (currentPhoto + depositAmount > limitPhoto) {
                printf("Le nombre dépasse la limite de %d. Opération annulée.\n", limitPhoto);
                return;
            }
            break;
        case 4: // CartesVisite
            int currentVisite = currentBalance.cartesVisite;
            int limitVisite = getCategoryLimit("CartesVisite");
            printf("Entrez le nombre de carte de visite à déposer (limite : %d) : ", limitVisite - currentVisite);
            scanf("%d", &depositAmount);

            if (currentVisite + depositAmount > limitVisite) {
                printf("Le nombre dépasse la limite de %d. Opération annulée.\n", limitVisite);
                return;
            }
            break;
        default:
            break;
    }

    // Enregistrer la transaction de dépôt
    char transaction[200];
    snprintf(transaction, sizeof(transaction), "Date=%s;Type=Dépôt;Compte=%s;Categorie=%s;Montant=%.2lf\n",
             getCurrentDate(), accountName, depositCategories[depositCategory], depositAmount);
    appendToFile(getTransactionFilePath(), transaction);

    printf("Dépôt effectué avec succès !\n");
}

void withdrawMoney() {
    char accountName[50];
    printf("Entrez le nom du compte : ");
    scanf("%s", accountName);

    // Demander la catégorie d'objet pour le retrait
    char* withdrawalCategories[] = {"Argent", "CarteBancaire", "Permis", "PhotosIdentite", "CartesVisite"};
    int withdrawalCategory;
    printf("Sélectionnez la catégorie pour le retrait : \n");
    for (int i = 0; i < sizeof(withdrawalCategories) / sizeof(withdrawalCategories[0]); i++) {
        printf("%d. %s\n", i + 1, withdrawalCategories[i]);
    }
    scanf("%d", &withdrawalCategory);
    withdrawalCategory--;

    if (withdrawalCategory == -1) {
        printf("Opération annulée.\n");
        return;
    }

    double withdrawalAmount;

    Balance currentBalance = getCurrentBalance(accountName);
    double currentAmount = 0;

    switch (withdrawalCategory) {
        case 0: // Argent
            currentAmount = currentBalance.argent;
            break;
        case 1: // CarteBancaire
            currentAmount = currentBalance.cartesBancaires;
            break;
        case 2: // Permis
            currentAmount = currentBalance.permisDeConduire;
            break;
        case 3: // PhotosIdentite
            currentAmount = currentBalance.photosIdentite;
            break;
        case 4: // CartesVisite
            currentAmount = currentBalance.cartesVisite;
            break;
        default:
            break;
    }

    double limit = getCategoryLimit(withdrawalCategories[withdrawalCategory]);

    do {
        // Demander le montant à retirer
        switch (withdrawalCategory) {
            case 0: // Argent
                printf("Entrez le montant de Argent à retirer : ");
                scanf("%lf", &withdrawalAmount);
                break;
            case 1: // CarteBancaire
                printf("Entrez le nombre de CarteBancaire à retirer : ");
                scanf("%lf", &withdrawalAmount);
                break;
            case 2: // Permis
                printf("Entrez le nombre de Permis à retirer : ");
                scanf("%lf", &withdrawalAmount);
                break;
            case 3: // PhotosIdentite
                printf("Entrez le nombre de PhotosIdentite à retirer : ");
                scanf("%lf", &withdrawalAmount);
                break;
            case 4: // CartesVisite
                printf("Entrez le nombre de CartesVisite à retirer : ");
                scanf("%lf", &withdrawalAmount);
                break;
            default:
                break;
        }

        // Vérifier si le montant à retirer dépasse la limite
        if (withdrawalAmount > limit) {
            printf("Le montant à retirer dépasse la limite de %.2lf. Entrez une nouvelle valeur.\n", limit);
        }
    } while (withdrawalAmount > currentAmount || withdrawalAmount > limit);

    // Enregistrer la transaction de retrait
    char transaction[200];
    snprintf(transaction, sizeof(transaction), "Date=%s;Type=Retrait;Compte=%s;Categorie=%s;Montant=%.2lf\n",
             getCurrentDate(), accountName, withdrawalCategories[withdrawalCategory], withdrawalAmount);
    appendToFile(getTransactionFilePath(), transaction);

    printf("Retrait effectué avec succès !\n");
}

void showTransactionHistory() {
    char accountName[50];
    printf("Entrez le nom du compte : ");
    scanf("%s", accountName);

    char transactions[1000];
    readFromFile(getTransactionFilePath(), transactions);

    // Vérifier si le compte existe avant d'afficher l'historique
    if (accountExists(transactions, accountName)) {
        // Les trois dernières transactions
        char* lastThreeTransactions[3];
        char* token = strtok(transactions, "\n");
        int transactionCount = 0;

        while (token != NULL) {
            if (transactionCount >= 3) {
                break;
            }
            lastThreeTransactions[transactionCount++] = token;
            token = strtok(NULL, "\n");
        }

        if (transactionCount > 0) {
            printf("Les trois dernières transactions :\n");

            for (int i = 0; i < transactionCount; i++) {
                char* transactionDetails = parseTransaction(lastThreeTransactions[i]);

                // Afficher les détails de la transaction en fonction du type
                char transactionType[MAX_CATEGORY_LENGTH];
                sscanf(transactionDetails, "Type=%s", transactionType);

                if (strcmp(transactionType, "Création") == 0) {
                    char argentDetails[MAX_CATEGORY_LENGTH];
                    sscanf(transactionDetails, "Argent=%s", argentDetails);
                    printf("- Type: %s, Montant initial: %s\n", transactionType, argentDetails);
                } else if (strcmp(transactionType, "Dépôt") == 0 || strcmp(transactionType, "Retrait") == 0) {
                    char categoryDetails[MAX_CATEGORY_LENGTH];
                    sscanf(transactionDetails, "Categorie=%s", categoryDetails);
                    char amountDetails[MAX_CATEGORY_LENGTH];
                    sscanf(transactionDetails, "Montant=%s", amountDetails);
                    printf("- Type: %s, Categorie: %s, Montant: %s\n", transactionType, categoryDetails, amountDetails);
                }
            }
        } else {
            printf("Aucune transaction effectuée jusqu'à présent.\n");
        }
    } else {
        printf("Le compte \"%s\" n'existe pas.\n", accountName);
    }
}

File: Transaction.cs
using System;
using System.IO;

class Transaction
{
    private readonly Utils _utils;

    public Transaction(Utils utils)
    {
        _utils = utils;
    }

    public void DepositMoney()
    {
        Console.Write("Entrez le nom du compte : ");
        string accountName = Console.ReadLine();

        Console.WriteLine("Sélectionnez la catégorie pour le dépôt : ");
        for (int i = 0; i < categories.Length; i++)
        {
            Console.WriteLine($"{i + 1}. {categories[i]}");
        }

        int depositCategory;
        while (!int.TryParse(Console.ReadLine(), out depositCategory) || depositCategory < 1 || depositCategory > categories.Length)
        {
            Console.WriteLine("Veuillez entrer une option valide.");
        }

        if (depositCategory == 0)
        {
            Console.WriteLine("Opération annulée.");
            return;
        }

        double depositAmount = 0;
        var currentBalance = _utils.GetCurrentBalance(accountName);

        switch (depositCategory - 1)
        {
            case 0: // Argent
                double currentArgent = currentBalance.Argent ?? 0;
                double limit = GetCategoryLimit("Argent");
                Console.Write($"Entrez le montant de Argent à déposer (limite: {limit - currentArgent}) : ");
                while (!double.TryParse(Console.ReadLine(), out depositAmount) || currentArgent + depositAmount > limit)
                {
                    Console.WriteLine($"Le montant total de Argent dépasse la limite de {limit}. Entrez une nouvelle valeur.");
                }
                break;
            // Ajoutez des cas pour les autres catégories ici
            default:
                break;
        }

        // Enregistrer la transaction de dépôt
        string transaction = $"Date={_utils.GetCurrentDate()};Type=Dépôt;Compte={accountName};Categorie={categories[depositCategory - 1]};Montant={depositAmount}\n";
        File.AppendAllText(_utils.GetTransactionFilePath(), transaction);

        Console.WriteLine("Dépôt effectué avec succès !");
    }

    public void WithdrawMoney()
        {
            Console.Write("Entrez le nom du compte : ");
            string accountName = Console.ReadLine();

            Console.WriteLine("Sélectionnez la catégorie pour le retrait : ");
            for (int i = 0; i < categories.Length; i++)
            {
                Console.WriteLine($"{i + 1}. {categories[i]}");
            }

            int withdrawalCategory;
            while (!int.TryParse(Console.ReadLine(), out withdrawalCategory) || withdrawalCategory < 1 || withdrawalCategory > categories.Length)
            {
                Console.WriteLine("Veuillez entrer une option valide.");
            }

            if (withdrawalCategory == 0)
            {
                Console.WriteLine("Opération annulée.");
                return;
            }

            double withdrawalAmount;

            var currentBalance = _utils.GetCurrentBalance(accountName);
            double currentAmount = currentBalance[categories[withdrawalCategory - 1]] ?? 0;
            double limit = GetCategoryLimit(categories[withdrawalCategory - 1]);

            do
            {
                // Demander le montant à retirer
                Console.Write($"Entrez le montant de {categories[withdrawalCategory - 1]} à retirer : ");
                while (!double.TryParse(Console.ReadLine(), out withdrawalAmount) || withdrawalAmount > limit)
                {
                    Console.WriteLine($"Le montant à retirer dépasse la limite de {limit}. Entrez une nouvelle valeur.");
                }
            } while (withdrawalAmount > currentAmount || withdrawalAmount > limit);

            // Enregistrer la transaction de retrait
            string transaction = $"Date={_utils.GetCurrentDate()};Type=Retrait;Compte={accountName};Categorie={categories[withdrawalCategory - 1]};Montant={withdrawalAmount}\n";
            File.AppendAllText(_utils.GetTransactionFilePath(), transaction);

            Console.WriteLine("Retrait effectué avec succès !");
        }

        public void ShowTransactionHistory()
        {
            Console.Write("Entrez le nom du compte : ");
            string accountName = Console.ReadLine();

            string transactions = File.ReadAllText(_utils.GetTransactionFilePath());

            // Vérifier si le compte existe avant d'afficher l'historique
            if (_utils.AccountExists(transactions, accountName))
            {
                // Les trois dernières transactions
                string[] lastThreeTransactions = transactions.Trim().Split('\n').Reverse().Take(3).Reverse().ToArray();

                if (lastThreeTransactions.Length > 0)
                {
                    Console.WriteLine("Les trois dernières transactions :");

                    foreach (string transaction in lastThreeTransactions)
                    {
                        var transactionDetails = _utils.ParseTransaction(transaction);

                        // Afficher les détails de la transaction en fonction du type
                        switch (transactionDetails.Type)
                        {
                            case "Création":
                                Console.WriteLine($"- Type: {transactionDetails.Type}, Montant initial: {transactionDetails.Argent}");
                                break;
                            case "Dépôt":
                            case "Retrait":
                                Console.WriteLine($"- Type: {transactionDetails.Type}, Categorie: {transactionDetails.Categorie}, Montant: {transactionDetails.Montant}");
                                break;
                            default:
                                break;
                        }
                    }
                }
                else
                {
                    Console.WriteLine("Aucune transaction effectuée jusqu'à présent.");
                }
            }
            else
            {
                Console.WriteLine($"Le compte \"{accountName}\" n'existe pas.");
            }
        }
}